### PR TITLE
fix: better handle concurrent builds

### LINKED ISSUE #3165: Header extraction is unreliable and results in truncated files
I frequently see corrupted node/electron headers. Sounds similar to #1054, but I'm using `npm`. And I see this on an isolate github actions runner, so it doesn't seem like a concurrency issue either?

```
npm error gyp info using node-gyp@10.1.0
npm error gyp info using node@20.18.2 | linux | x64
npm error gyp info find Python using Python version 3.12.3 found at "/usr/bin/python3"
npm error gyp http GET https://electronjs-mirror.anysphere.workers.dev/headers/v34.5.1/node-v34.5.1-headers.tar.gz
npm error gyp http 200 https://electronjs-mirror.anysphere.workers.dev/headers/v34.5.1/node-v34.5.1-headers.tar.gz
npm error gyp http GET https://electronjs-mirror.anysphere.workers.dev/headers/v34.5.1/SHASUMS[25](https://github.com/anysphere/everysphere/actions/runs/15405526830/job/43347404983#step:14:26)6.txt
npm error gyp http 200 https://electronjs-mirror.anysphere.workers.dev/headers/v34.5.1/SHASUMS256.txt
...
npm error   File "/home/runner/.cache/node-gyp/34.5.1/include/node/common.gypi", line 177
npm error     'msvs_settings': {
npm error                      ^
npm error SyntaxError: '{' was never closed
```

more examples:

```
npm error /home/runner/.cache/node-gyp/34.3.4/include/node/v8-object.h:5: error: unterminated #ifndef
npm error     5 | #ifndef INCLUDE_V8_OBJECT_H_
```

```
npm error gyp: /Users/ec2-user/Library/Caches/node-gyp/20.18.2/common.gypi not found while reading includes of binding.gyp while trying to load binding.gyp
```

```
npm error gyp: /home/AzDevOps/.cache/node-gyp/30.5.1/common.gypi not found while reading includes of binding.gyp while trying to load binding.gyp
```

We see this on both linux and macos, but interestingly never on windows. Perhaps due to this?

https://github.com/nodejs/node-gyp/blob/7d883b5cf4c26e76065201f85b0be36d5ebdcc0e/lib/install.js#L203-L206

I would propose making the separate directory download step the default on all platforms.

### PR DESCRIPTION
**Pull request by dsanders11**
##### Checklist

- [x] `npm install && npm run lint && npm test` passes
- [x] tests are included 
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)

##### Description of change

Fixes #3165, which unfortunately was not actually fixed by #3170. I'm not sure #3170 had any effect on the issue, as my original changes in #2846 which were extended to all platforms in #3710 were targeting a specific issue on Windows, and did not aim to fix (or know of) the issue in #3165.

The second commit on this branch adds a test which repros the issue being reported in #3165, namely that header files can be seen as partially written when multiple builds are happening concurrently. A failing test run before the other commits on this branch [can be seen here](https://github.com/dsanders11/node-gyp/actions/runs/24318080696).

The remainder of the commits refactor `copyDirectory` to fix the issue, fix another concurrent build issue I came across during testing, and adds more test coverage. This refactor also allows for dropping `exponential-backoff` as a dependency.

### CODE REVIEW

File: test/test-copy-directory.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        polls++
+        if (stat.size !== fileSize) {
+          violations.push({ poll: polls, size: stat.size })
+        }
+      } catch (err) {
+        if (err.code !== 'ENOENT') throw err
+      }
+    }, 0)
+
+    await copyDirectory(srcDir, destDir)
```
Comment (legendecas) [inline test/test-copy-directory.js]: Would you mind adding a test that `srcDir` is unchanged?

File: lib/copy-directory.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    // temp path in the dest directory first, then rename within the
+    // same filesystem to keep it atomic.
+    //
+    // When another process wins the race, rename may fail with one of
+    // these codes — all mean the destination was already placed and
+    // are safe to ignore since every process extracts identical content.
+    const srcPath = path.join(src, entry.name)
+    const destPath = path.join(dest, entry.name)
+    try {
+      await fs.rename(srcPath, destPath)
```
Comment (legendecas) [inline lib/copy-directory.js:34]: This changes the semantics to `moveDirectory` instead of `copyDirectory`.

Comment (dsanders11) [reply lib/copy-directory.js:34]: Happy to rename it and the files. You're right that now it's move, outside of the edge case of cross-device on Windows, which falls back to a copy and move.

File: test/test-copy-directory.js (hunk 1 above)
Comment (dsanders11) [reply test/test-copy-directory.js]: I've added a test that `moveDirectory` (renamed it to reflect the changed semantics) retains the contents from `srcDir` moved to `destDir`. Let me know if I misunderstood what you were asking for here and I can add more test coverage.

File: lib/copy-directory.js (hunk 2 above)
Comment (oga35767-eng) [reply lib/copy-directory.js:34]: 👍

**Review by oga35767-eng — APPROVED** [review]

**Comment by oga35767-eng** [comment]
👍

File: lib/copy-directory.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    } catch (err) {
+      if (RACE_ERRORS.includes(err.code)) {
+        // Another parallel process already placed this entry — ignore
```
Comment (legendecas) [inline lib/copy-directory.js:37]: I don't think this is a really good idea. This could result in a folder where some of its sub-entries are stale (like on windows it's still been used resulting in EBUSY), but some are new.

A directory already at the destination is not proof the contents match.
